solana/solana-ibc/programs/solana-ibc/src/transfer/impls.rs

```diff
@@ -1,26 +1,209 @@
+        });
+
+        let amount = amt.amount;
+        let amount_in_u64 = check_amount_overflow(amount)?;
+
+        let (_token_mint_key, _bump) =
+            Pubkey::find_program_address(&[base_denom.as_ref()], &crate::ID);
+        let store = self.borrow();
+        let accounts = &store.accounts;
+
+        let sender = get_account_info_from_key(accounts, sender_id)?;
+        let receiver = get_account_info_from_key(accounts, receiver_id)?;
+        let token_program = get_account_info_from_key(accounts, spl_token::ID)?;
+
+        if matches!(from, AccountIdx::Escrow(_)) {
+            let (mint_authority_key, mint_authority_bump) =
+                Pubkey::find_program_address(&[MINT_ESCROW_SEED], &crate::ID);
+
+            let mint_authority =
+                get_account_info_from_key(accounts, mint_authority_key)?;
+
+            let bump_vector = mint_authority_bump.to_le_bytes();
+            let inner = vec![MINT_ESCROW_SEED, bump_vector.as_ref()];
+            let outer = vec![inner.as_slice()];
+
+            // Below is the actual instruction that we are going to send to the Token program.
+            let transfer_instruction = Transfer {
+                from: sender.clone(),
+                to: receiver.clone(),
+                authority: mint_authority.clone(),
+            };
+            let cpi_ctx = CpiContext::new_with_signer(
+                token_program.clone(),
+                transfer_instruction,
+                outer.as_slice(), //signer PDA
+            );
+
+            anchor_spl::token::transfer(cpi_ctx, amount_in_u64).unwrap();
+        } else {
+            let sender_token_account =
+                TokenAccount::try_deserialize(&mut &sender.data.borrow()[..])
+                    .unwrap();
+            let sender_token_account_owner = sender_token_account.owner;
+            let authority = get_account_info_from_key(
+                accounts,
+                sender_token_account_owner,
+            )?;
+
+            // PDA generated so that we can sign the tx
+            let (_mint_authority_key, mint_authority_bump) =
+                Pubkey::find_program_address(&[MINT_ESCROW_SEED], &crate::ID);
+            let bump_vector = mint_authority_bump.to_le_bytes();
+            let inner = vec![MINT_ESCROW_SEED, bump_vector.as_ref()];
+            let outer = vec![inner.as_slice()];
+
+            // Below is the actual instruction that we are going to send to the Token program.
+            let transfer_instruction = Transfer {
+                from: sender.clone(),
+                to: receiver.clone(),
+                authority: authority.clone(),
+            };
+            let cpi_ctx = CpiContext::new_with_signer(
+                token_program.clone(),
+                transfer_instruction,
+                outer.as_slice(), //signer PDA
+            );
+
+            anchor_spl::token::transfer(cpi_ctx, amount_in_u64).unwrap();
+        }
```

Comment:
There seems to be quite a bit of common code between both branches.  Isn’t this correct:

```suggestion
        let (mint_authority_key, mint_authority_bump) =
            Pubkey::find_program_address(&[MINT_ESCROW_SEED], &crate::ID);

        let seeds = [MINT_ESCROW_SEED, core::slice::from_ref(&mint_authority_bump)];
        let seeds = seeds.as_ref();
        let seeds = core::slice::from_ref(&seeds);

        let authority = if matches!(from, AccountIdx::Escrow(_)) {
            get_account_info_from_key(accounts, mint_authority_key)?
        } else {
            let sender_token_account =
                TokenAccount::try_from_slice(&sender.data.borrow()[..])?;
            get_account_info_from_key(accounts, sender_token_account.owner)?
        }

        let transfer_instruction = Transfer {
            from: sender.clone(),
            to: receiver.clone(),
            authority: authority.clone(),
        };
        let cpi_ctx = CpiContext::new_with_signer(
            token_program.clone(),
            transfer_instruction,
            seeds,
        );
        anchor_spl::token::transfer(cpi_ctx, amount)?;
```

Also, probably should prefer returning errors to `unwrap`.